### PR TITLE
ABI Cleanup

### PR DESCRIPTION
**Pull request by barakman**
Handles issue #415 partially:
1. Rewrite every ABI as a dictionary declared in a properly-formatted manner, rather than as a (very long) string which is json-loaded into a dictionary during runtime
2. Keep in every ABI only the functions and events that the bot relies on

### CODE REVIEW

**Review by mikewcasale — DISMISSED** [review]

**Review by sklbancor — DISMISSED** [review]
overall this seems like a trivial change; assuming tests pass [and fingers crossed that our tests cover all ABI use cases and we are not missing some ABI here] I don't see any reasons to not merge it

**Comment by barakman** [comment]
> overall this seems like a trivial change; assuming tests pass [and fingers crossed that our tests cover all ABI use cases and we are not missing some ABI here] I don't see any reasons to not merge it

Trivial though pretty difficult to complete.
I've actually implemented a couple of scripts in order to achieve that:
1. A script which re-parses this file and changes its contents from JSON strings into JSON objects
2. A script which scans the entire code-base, finds all functions and events which are potentially being used, and removes all the rest

**Comment by sklbancor** [comment]
> > overall this seems like a trivial change; assuming tests pass [and fingers crossed that our tests cover all ABI use cases and we are not missing some ABI here] I don't see any reasons to not merge it
> 
> Trivial though pretty difficult to complete. I've actually implemented a couple of scripts in order to achieve that:
> 
> 1. A script which re-parses this file and changes its contents from JSON strings into JSON objects
> 2. A script which scans the entire code-base, finds all functions and events which are potentially being used, and removes all the rest from those JSON objects

Jupyter actually allows you to do No 1 -- albeit possibly not in the size required. But with the snippet

```
data = json.loads(data_s)
data
```
you produce output that then you can import into Python

**Comment by barakman** [comment]
> > > overall this seems like a trivial change; assuming tests pass [and fingers crossed that our tests cover all ABI use cases and we are not missing some ABI here] I don't see any reasons to not merge it
> > 
> > 
> > Trivial though pretty difficult to complete. I've actually implemented a couple of scripts in order to achieve that:
> > 
> > 1. A script which re-parses this file and changes its contents from JSON strings into JSON objects
> > 2. A script which scans the entire code-base, finds all functions and events which are potentially being used, and removes all the rest from those JSON objects
> 
> Jupyter actually allows you to do No 1 -- albeit possibly not in the size required. But with the snippet
> 
> ```
> data = json.loads(data_s)
> data
> ```
> 
> you produce output that then you can import into Python

That's pure Python, no need for Jupyter.

But that's not as simple, since I wanted to format the primary fields with a standard indentation of 4 spaces, and the internal fields (the values of 'input' and 'output') in a single line.

A kind of a mitigation between a readable file and a relatively small file, since applying the indentation for everything would have resulted with a huge file, and not applying it at all would have resulted with an unreadable file.

Also, I wanted to rearrange the fields of each element, starting with type ("function" or "event"), then name, and so on, since the Solidity compiler arranges them in a pretty weird order IMO.

So in short, a script was required hee as well.

**Review by sklbancor — APPROVED** [review]

File: fastlane_bot/tests/test_037_Exchanges.py [hunk 1]
```diff
@@ -174,7 +173,7 @@ def test_test_uniswap_v2_exchange_fork():
     async def test_uniswap_v2_exchange_fork():
```
Comment (sklbancor) [inline fastlane_bot/tests/test_037_Exchanges.py:173]: Unrelated to this PR but I need this comment to make it another issue

```
async def test_uniswap_v2_exchange_fork():
        assert (pancake_v2_exchange.exchange_name in "pancakeswap_v2"), f"Wrong exchange name. Expected pancakeswap_v2, got {pancake_v2_exchange.exchange_name}"
        assert (pancake_v2_exchange.base_exchange_name in "uniswap_v2"), f"Wrong base exchange name. Expected uniswap_v2, got {pancake_v2_exchange.base_exchange_name}"  
```

The whole point of defining constants is that they abstract from the underlying values. The only reason why we define

`PANCAKESWAP_V2_NAME = "pancakeswap_v2"`

is because this makes debugging easier. However, if at one point we would like to change 

`PANCAKESWAP_V2_NAME = 1234`

then the code should still work as previously.

Relatedly -- tests are also contracts with the users of the library. If you do 

`assert PANCAKESWAP_V2_NAME == "pancakeswap_v2"`

then this means that the user of the library can rely on this association being true (and in SemVer this means that if ever you can it you need to change the major version number)

**Review by sklbancor — APPROVED** [review]
